addons/tls13-middlebox/bootstrap.js

```diff
@@ -0,0 +1,241 @@
+"use strict";
+
+let {classes: Cc, interfaces: Ci, utils: Cu, results: Cr} = Components;
+
+Cu.import("resource://gre/modules/Preferences.jsm");
+Cu.import("resource://gre/modules/TelemetryController.jsm");
+Cu.import("resource://gre/modules/XPCOMUtils.jsm");
+
+const VERSION_MAX_PREF = "security.tls.version.max";
+const FALLBACK_LIMIT_PREF = "security.tls.version.fallback-limit";
+
+let readwrite_prefs = new Preferences({defaultBranch: true});
+
+// all combination of configurations we care about.
+let configurations = [
+  {maxVersion: 4, fallbackLimit: 4, website: "enabled.tls13.com"},
+  {maxVersion: 4, fallbackLimit: 4, website: "disabled.tls13.com"},
+  {maxVersion: 3, fallbackLimit: 3, website: "control.tls12.com"}
+];
+
+// some fields are not available sometimes, so we have to catch the errors and return undefined.
+function getFieldValue(obj, name) {
+  try {
+    return obj[name];
+  } catch (ex) {
+    return undefined;
+  }
+}
+
+// extracting the root certificate from the chain
+function extractRootCert(cert) {
+  let root_cert = cert;
+
+  while (getFieldValue(root_cert, "issuer")) {
+    root_cert = getFieldValue(root_cert, "issuer");
+  }
+
+  return root_cert;
+}
+
+// returns true if there is at least one non-builtin root certificate is installed
+function isNonBuiltInRootCertInstalled() {
+  let certDB = Cc["@mozilla.org/security/x509certdb;1"].getService(Ci.nsIX509CertDB);
+
+  let iter = certDB.getCerts().getEnumerator();
+
+  while (iter.hasMoreElements()) {
+    let cert = iter.getNext().QueryInterface(Ci.nsIX509Cert);
+
+    if (getFieldValue(cert, "issuer") === null &&
+      getFieldValue(cert, "isBuiltInRoot") === false &&
+      // There are some root certificates that are built-in but their isBuiltInRoot is false.
+      // That is why we check their tokenName as well
+      getFieldValue(cert, "tokenName").toLowerCase() !== "Builtin Object Token".toLowerCase()) {
+      return true;
```

Comment:
You could just build a list of the expected built-in certificates (by hash or similar) and see if nsIX509CertDB.getCerts() returns any not in that list.